Something amiss with VimL parser (missing :endif in plugin)

- `nvim --version`:  NVIM v0.2.0-1161-g6afa7d66
- Vim (version: ) behaves differently? Yes, it doesn't error

VIM - Vi IMproved 8.0 (2016 Sep 12, compiled Mar 07 2017 03:33:23)
Included patches: 1-197, 322, 377-378

- Operating system/version: Linux
- Terminal name/version: urxvt
- `$TERM`: rxvt-unicode

### Actual behaviour

nvim prints the following before painting the editor window. This seemingly has to do with https://github.com/xolox/vim-misc which is required by https://github.com/xolox/vim-easytags. The error has not been printed by vanilla Vim up until now. Sadly, it's not 100% reproducible here (as I just noticed, after collecting a lot of info). I'm still keeping the text I gathered around, so ti can be referred to later.

```
Error detected while processing function xolox#easytags#initialize[33]..xolox#misc#path#which:
line   25:
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
line   34:
E580: :endif without :if:       endif
line   25:
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
E171: Missing :endif:         endfor
line   34:
E580: :endif without :if:       endif
line   25:
E171: Missing :endif:         endfor
line   34:
E580: :endif without :if:       endif
```

The function in question is: https://github.com/xolox/vim-misc/blob/3e6b8fb6f03f13434543ce1f5d24f6a5d3f34f0b/autoload/xolox/misc/path.vim#L9-L47. Which actually seems well-formed.

Since I can't figure out what's wrong with the function that nvim thinks, I don't know how to reproduce this with nvim -u NORC. I noticed this problem because I just rebuilt nvim, my last rebuild was from about ~2 weeks ago. While bisecting, the majority of commits fail to build with this message:

```
/home/aktau/github/neovim/neovim/src/nvim/eval.c:7984:35: error: macro "strncmp" requires 3 arguments, but only 2 given
     if (strncmp(s, S_LEN("silent")) == 0) {
                                   ^
```

(BTW, I thought strncmp(3) wasn't a macro.)

At the point in the tree where this particular error has been fixed (29bad04f I think), another error comes into play: 

```
/home/aktau/github/neovim/neovim/src/nvim/eval.c: At top
level:
cc1: warning: unrecognized command line option "-Wno-static-in-inline"
[enabled by default]
[100%] Building C object src/nvim/CMakeFiles/nvim.dir/ui.c.o
[100%] Building C object src/nvim/CMakeFiles/nvim.dir/ui_bridge.c.o
[100%] Building C object src/nvim/CMakeFiles/nvim.dir/undo.c.o
[100%] Building C object src/nvim/CMakeFiles/nvim.dir/version.c.o
[100%] Building C object src/nvim/CMakeFiles/nvim.dir/window.c.o
[100%] Linking C executable ../../bin/nvim
CMakeFiles/nvim.dir/eval.c.o: In function `f_getcompletion':
/home/aktau/github/neovim/neovim/src/nvim/eval.c:9465:
undefined reference to `get_tv_string'
/home/aktau/github/neovim/neovim/src/nvim/eval.c:9466:
undefined reference to `get_tv_string'
collect2: error: ld returned 1 exit status
make[3]: *** [bin/nvim] Error 1
make[3]: Leaving directory
`/home/aktau/github/neovim/neovim/build'
make[2]: *** [src/nvim/CMakeFiles/nvim.dir/all] Error 2
make[2]: Leaving directory
`/home/aktau/github/neovim/neovim/build'
make[1]: *** [all] Error 2
make[1]: Leaving directory
`/home/aktau/github/neovim/neovim/build'
make: *** [nvim] Error 2
```

Which makes the bisect take a very long time (lots of `git bisect skip`). And (I think) loop back upon itself, making it impossible to find the commit that caused this. I suspect @ZyX-I might know which one is the exact culprit.

After a couple of these loops, I got tired of it and saved the git bisect state:

```
05b74399  2017-03-29  Justin M. Keyes                                                                             build: remove unused get_preproc_output() call
2846d508  2017-03-29  lonerover                                                                                   vim-patch:7.4.2276 (#6393)
46efe144  2017-03-28  ZyX                                                                                         functests: Try sleeping a bit more
b9603218  2017-03-28  ZyX                                                                                         eval/executor: Fix check-single-includes
a1d590a0  2017-03-28  ZyX                              refs/bisect/bad                                            *: Use const char * in set_one_cmd_context
114eaa15  2017-03-28  ZyX                              refs/bisect/skip-114eaa15f009cbd3e3deb177a2a67affa430fbb8  eval/typval,api/buffer: Fix review comments
58e34e8d  2017-03-26  ZyX                                                                                         eval/typval: Allow NULL dict as tv_dict_get_callback() argument
f4256243  2017-03-25  ZyX                              refs/bisect/skip-f4256243dbdbd2858be716f9f52763bd9a76be4d  eval: Fix -Werror=unitialized from QB
29bad04f  2017-03-25  ZyX                              refs/bisect/skip-29bad04f9e2e70b7c1c198ca0d86d23282bee8a1  eval: Do not supply S_LEN to strncmp
8daf756f  2017-03-25  ZyX                              refs/bisect/skip-8daf756fb6f92bdeb39e473b34364afd5270dd99  unittests: Fix linter errors
7826ee1c  2017-03-25  ZyX                                                                                         unittests: Add tv_get_string* tests
e08b27ba  2017-03-25  ZyX                              HEAD                                                       unittests: Add tv_get number tests
4536c064  2017-03-25  ZyX                                                                                         unittests: Move tv_dict_add* tests to a proper describe() block
49195063  2017-03-25  ZyX                                                                                         unittests: Add tv_check… tests
389274be  2017-03-25  ZyX                                                                                         unittests: Add tv_equal() tests
630ff33d  2017-03-25  ZyX                                                                                         unittests: Test locks section
ed4948a9  2017-03-13  ZyX                                                                                         unittests: Test tv_copy()
f0bbd1e8  2017-03-13  ZyX                                                                                         unittests: Add tests for tv_clear()
e43de6bb  2017-03-13  ZyX                                                                                         unittests: Add test for tv_dict_set_keys_readonly
368a61c5  2017-03-13  ZyX                                                                                         unittests: Add tv_dict_copy tests
218fa1d8  2017-03-13  ZyX                              refs/bisect/skip-218fa1d8069c1f1d8c5154b1abc2b686adc0c742  charset: Remove useless condition from vim_iswordc_tab
8b9a1fbf  2017-03-13  ZyX                              refs/bisect/skip-8b9a1fbf7a630b68b1428a39f25e1fa38fe0cc9f  unittests: Add tests for tv_dict_extend
fa852e7c  2017-03-13  ZyX                              refs/bisect/skip-fa852e7cdc365b6fcd39d677f4067963274c44c3  eval: Fix extend() behaviour with NULL lists and dictionaries
4987850c  2017-03-13  ZyX                              refs/bisect/skip-4987850cacc9f0c5674bc4cb8197a6fa0a343167  unittests: Add tv_dict_clear tests
43e9fad1  2017-03-13  ZyX                              refs/bisect/skip-43e9fad1c8835a3136f4db53c82608e034df2a5e  eval: Use tv_is_func in place of ==VAR_FUNC                                                                                                                                                                                                                                                                                                        ==VAR_PARTIAL
270a3889  2017-03-13  ZyX                              refs/bisect/skip-270a3889af024485fa7b63f34c4dd3f92f6e0f98  unittests: Add tv_dict_add* unit tests
bc87d23c  2017-03-12  ZyX                              refs/bisect/skip-bc87d23c28c10c70b4addaf18ae16bcbe0682c8a  unittests: Add tests for dictionary indexing
b222453c  2017-03-12  ZyX                              refs/bisect/skip-b222453c95dbe466b79abd578642fa6038770a55  eval/typval: Refactor errors a bit: use emsgf always
5ce62432  2017-03-12  ZyX                              refs/bisect/skip-5ce624324136408a3f412d6c2ff437455f1e3344  unittests: Enable tv_list_join tests back
52e226ff  2017-03-05  ZyX                              refs/bisect/skip-52e226ff74f49a02f884247b61d1865fa14e810d  unittests: Disable tv_list_join test on Mac OS only
f830243f  2017-03-05  ZyX                              refs/bisect/skip-f830243ff7b78b9bb41e72fd602b99c2893b5437  mbyte: Include os_defs.h in mbyte.h
3bf87a5a  2017-03-05  ZyX                                                                                         eval: Do not use S_LEN as snprintf argument
38dd81c1  2017-03-05  ZyX                              refs/bisect/skip-38dd81c136c2edbda66614622a4747bf1785af94  eval/typval: Fix SEGV in test_alot.vim test
faddd83d  2017-03-05  ZyX                                                                                         eval: Fix SEGV in test49
6c622ed0  2017-03-05  ZyX                              refs/bisect/skip-6c622ed08bb56f08f4cc4dbf6251220ccf7ba2ba  unittests: Add tv_dict_item_{add,remove} tests
ffaf7c75  2017-03-05  ZyX                              refs/bisect/skip-ffaf7c7521399826d9a32b12a1180bcd162f88be  unittests: Add tv_dict_item_{alloc,free} tests
cdb1aa3e  2017-03-05  ZyX                              refs/bisect/skip-cdb1aa3e47cb0ec19d2ae597c1d21b7e892d0d7e  eval: Fix len argument to xstrlcat
2c8ad276  2017-03-05  ZyX                              refs/bisect/skip-2c8ad276523489223d2d804075a19367280b0f10  ascii: Readd DEL_STR define
c6c48e86  2017-03-04  ZyX                              refs/bisect/skip-c6c48e8672ab6adbf78453e2b133fe3bd2ee1eb5  syntax: Fix linter error
78a0de2c  2017-03-04  ZyX                              refs/bisect/skip-78a0de2c1b722d5445dfc44cc0635e5ce4bcd23e  eval/typval: Fix -Werror=return-type
4c3be98d  2017-03-04  ZyX                              refs/bisect/skip-4c3be98db9b24b7d5b8dc2cf55574e778b7ad137  unittests: Add tv_dict_watcher_{add,remove} tests
14017466  2017-03-04  ZyX                              refs/bisect/skip-140174669e359ae7147288366dd94952955980cc  unittests: Run tv_list_join tests in case it stopped failing
6aa6e500  2017-03-04  ZyX                              refs/bisect/skip-6aa6e5007596f1eaa23c45376e3a72d908c6688e  eval: Fix linter errors
506b9389  2017-02-14  ZyX                              refs/bisect/skip-506b938947b7083fc8bef20eabc08ed033298add  *: Make some more things const and with length
3025431c  2016-12-04  ZyX                              refs/bisect/skip-3025431c81daac873e69a71aee695ebfd00504f7  eval: Make sure that v:_null_dict does not crash dictwatcher*()
a56f2d27  2016-12-04  ZyX                              refs/bisect/skip-a56f2d27e3c09aaae00a58a70652ac5db3287dee  eval: Make dictionary watchers work with empty keys
1e3e302d  2016-12-04  ZyX                              refs/bisect/skip-1e3e302dc2bdced563ecd2c3fb101af31f72b3df  eval: Move part of dictwatcher* functions to eval/typval
52396162  2016-11-05  ZyX                              refs/bisect/skip-5239616297182601a5d244a482e6cef307e901f1  functests: Fix buf_functions test on Windows
4bcee963  2016-11-05  ZyX                              refs/bisect/skip-4bcee963471abd939bb9edd1709418e30be7290f  *: Fix some Windows-specific warnings
9ed9af7e  2016-11-05  ZyX                              refs/bisect/skip-9ed9af7e11e3a707f65abfeb1d02b029e39241ea  eval/typval: More `const` qualifiers in `tv_dict*` function signatures
56e51033  2016-11-05  ZyX                              refs/bisect/skip-56e51033abf00d66e9c6f9412e8f57c9a24b86ae  unittests: Add tests for tv_list_idx_of_item
e5edf07e  2016-11-04  ZyX                              refs/bisect/skip-e5edf07ec44f8d147d7482cae2997be62c30373f  unittests: Add tests for tv_list_find*() functions
b3672ae2  2016-10-02  ZyX                              refs/bisect/skip-b3672ae2fced4715963442d2e19048f8fadbe0b8  eval/typval: Add tv_list_equal() tests, compare NULL lists equal
4f9e7844  2016-09-25  ZyX                              refs/bisect/skip-4f9e7844272335943c075b9444934f77d3cba234  unittests: Test tv_list_join()
cf45c7bb  2016-09-25  ZyX                              refs/bisect/skip-cf45c7bb059dafeb882133273f77042cb06b37e4  unittests: Fix tests crash
7ceebacb  2016-09-24  ZyX                                                                                         eval/typval,tests: Fix extending list with itself, add tests
56e4c2f6  2016-09-19  ZyX                                                                                         unittests: Test tv_list_concat()
f80a0046  2016-09-18  ZyX                              refs/bisect/skip-f80a00469fdba8a3dec0edae30c911d050485055  eval/typval: Make tv_list_concat handle NULL lists correctly
82e6cac5  2016-09-18  ZyX                              refs/bisect/skip-82e6cac5f9d437da5aeb8c4cf51945cfeb5922ff  functests: Add null_spec.lua from #4615
9898f36a  2016-09-17  ZyX                              refs/bisect/skip-9898f36aa306d2a7f53fbe08c64048260992d3bb  unittests: Test tv_list_copy
9b8beaff  2016-09-17  ZyX                                                                                         unittests: Add tests for tv_list_insert*()/…append*() functions
be360d88  2016-09-14  ZyX                              refs/bisect/skip-be360d88414fa3b874262f3ed3a2bba5926a751e  unittests: Add tests for tv_list_insert()
d2639e1e  2016-09-14  ZyX                                                                                         unittests: Add tests for list watchers and list alloc/free/unref
a3941671  2017-03-12  ZyX                              refs/bisect/skip-a394167177390a66c275d32e3862c82c546970ff  unittests: Test tv_list_item_\* functions
a32db8ed  2016-09-11  ZyX                              refs/bisect/skip-a32db8ed19c35847373d4a7fd56d7797a5a26897  eval/typval: Add missing includes, also add a script to find them
2ad4fba4  2016-09-11  ZyX                              refs/bisect/skip-2ad4fba46db26d0724106b194f7cb628fb9b02e8  eval: Move copy_tv to eval/typval
8b0fa64e  2016-09-10  ZyX                              refs/bisect/skip-8b0fa64ed3dbd4376cbfa195526ec97d8c944a97  message: Remove incorrect assertion
40feac6e  2016-09-10  ZyX                                                                                         message: Revise maxlen argument in msg_puts_attr_len
3a3816c9  2016-09-04  ZyX                              refs/bisect/skip-3a3816c990362e1ad8321e6b4748a33c94d1d797  cmake: Use CMAKE_CURRENT_LIST_DIR and remove vars used only once
86fc4580  2016-09-04  ZyX                              refs/bisect/skip-86fc4580b83f20211d7f85566e84a0e1dad0a136  eval: Fix max_min functions
c4fe656f  2016-09-04  ZyX                              refs/bisect/skip-c4fe656fef5ade20da4861a2f1e5d0f776b0df8f  typval.h: Allow non-var expressions in TV_DICT_ITER first argument
6cc3d59e  2016-09-04  ZyX                              refs/bisect/skip-6cc3d59ec8e4d6f32c8c3d9755c625e32512b8e2  misc1: Refactor ask_yesno()
31a3158d  2016-09-04  ZyX                              refs/bisect/skip-31a3158d0b574385186ab55c074edf85356d1d6c  eval: Make sort always stable
5df35297  2016-09-04  ZyX                              refs/bisect/skip-5df35297f832b3247c18253c916be6066c603739  eval: Remove eval_expr() completely
c8e63a8d  2016-09-04  ZyX                              refs/bisect/skip-c8e63a8db84e9d9f7bd855085a87d93631504fc7  eval: Move remaining get_tv_string* functions to eval/typval.c
50ebd1df  2016-08-28  ZyX                              refs/bisect/skip-50ebd1dff5c4e995c4f7e7980870e43d9defabc6  eval: Move free_tv to eval/typval.h, remove most of its usages
233b0c93  2016-08-28  ZyX                                                                                         eval: Move get_tv_number[_chk] to eval/typval.c
1b3e13da  2016-08-28  ZyX                              refs/bisect/skip-1b3e13da5be48739ac4291208e421b68e607cc0f  eval: Refactor get_tv_lnum_buf
7ee5cc74  2016-08-28  ZyX                              refs/bisect/skip-7ee5cc7429017a4a08a8b62b628bea156fea0008  eval: Move get_tv_lnum and get_tv_float to eval/typval.h
949f09bd  2016-08-21  ZyX                                                                                         eval: Move get_tv_string_buf() to eval/typval.c
28dafe3f  2016-08-21  ZyX                              refs/bisect/skip-28dafe3ff0b0dc082fb62b2251fd64a167ce7188  eval,*: Move get_tv_string to typval.c
5cdf7177  2016-08-21  ZyX                                                                                         eval: Move get_float_arg to typval.h
2dcfc439  2016-08-20  ZyX                                                                                         eval: Split and move dict_add_nr_str to typval.c
210342d7  2017-01-11  ZyX                              refs/bisect/good-210342d795a13e88579d3e55ae931463d16f241d  eval: Move dict_add_list and dict_add_dict to typval.c
983a5532  2016-08-20  ZyX                              refs/bisect/good-983a5532ca17060508ea308a9b7069dae4facb59  eval: Move dict_set_keys_readonly to typval.c
ecff8387  2016-08-20  ZyX                                                                                         eval: Move get_dict_callback to typval.c
54bd2e8b  2016-08-20  ZyX                                                                                         eval: Make setmatches() return -1 in case of some failures
e18a5783  2016-08-20  ZyX                              refs/bisect/good-e18a5783080f7c94f408ec5f53dedffdb69789e1  *: Move some dictionary functions to typval.h and use char*
50a48f2a  2016-07-30  ZyX                              refs/bisect/good-50a48f2a0ecf7f767df961f7f5060505cf28e331  functests: Add tests for some *buf* functions
fb146e80  2016-07-26  ZyX                                                                                         eval: Split eval.c into smaller files
18e7d552  2017-03-29  Justin M. Keyes                                                                             terminal.c:redraw(): Avoid invalid cursor col (#6379)
e86042ae  2017-03-29  lonerover                                                                                   vim-patch:7.4.2343 and mark NA patches (#6384)
b9e7ab14  2017-03-28  Samuel Catherasoo                                                                           refactor/single-include: charset.h (#6385)
```

The ones with `skip-` refs don't compile for some of the reasons listed above. The culprit is somewhere in there, but I'd need to make patches for the compilation issues and apply them during bisection. Keeping this bug to pick it up when I get time.

Hmmm, it doesn't find the packages apparently.

```
$  rm -rf build .deps ; ( mkcd build && brewless env CC=clang-3.8 CXX=clang++-3.8 cmake -DCLANG_ASAN_UBSAN=ON -DCMAKE_INSTALL_PREFIX:PATH=$HOME/neovim .. ) && make
-- The C compiler identification is Clang 3.8.0
-- The CXX compiler identification is Clang 3.8.0
-- Check for working C compiler: /usr/bin/clang-3.8
-- Check for working C compiler: /usr/bin/clang-3.8 -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working CXX compiler: /usr/bin/clang++-3.8
-- Check for working CXX compiler: /usr/bin/clang++-3.8 -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- CMAKE_BUILD_TYPE not given, defaulting to 'Dev'.
-- Replacing -O3 in CMAKE_C_FLAGS_RELEASE with -O2.
-- Performing Test HAS_ACCEPTABLE_FORTIFY
-- Performing Test HAS_ACCEPTABLE_FORTIFY - Success
...
-- Found PkgConfig: /usr/bin/pkg-config (found version "0.26") 
...
CMake Error at /usr/share/cmake-2.8/Modules/FindPackageHandleStandardArgs.cmake:108 (message):
  Could NOT find LibUV (missing: LIBUV_LIBRARY LIBUV_INCLUDE_DIR)
Call Stack (most recent call first):
  /usr/share/cmake-2.8/Modules/FindPackageHandleStandardArgs.cmake:315 (_FPHSA_FAILURE_MESSAGE)
  cmake/FindLibUV.cmake:105 (find_package_handle_standard_args)
  CMakeLists.txt:307 (find_package)
```

brewless is (similar result with and without brewless):

```
brewless () {
	(
		env | grep linuxbrew | sed -r -e 's@([=:])[^:]*.linuxbrew[^:]*:@\1@' | while read i
		do
			export $i
		done
		"$@"
	)
}
```

My normal command is:

```
 make distclean && make -j12 CMAKE_BUILD_TYPE=RelWithDebInfo CMAKE_EXTRA_FLAGS="-DCMAKE_INSTALL_PREFIX:PATH=$HOME/neovim" install
```

Which works. So I need (or someone more experienced) needs to figure out what the difference is.

--- a/src/nvim/eval.c
+++ b/src/nvim/eval.c
@@ -13416,14 +13416,12 @@ static void f_resolve(typval_T *argvars, typval_T *rettv, FunPtr fptr)
           q = (char *)path_tail((char_u *)p);
         }
         if (q > p && !path_is_absolute_path((const char_u *)buf)) {
-          // Symlink is relative to directory of argument.
+          // Symlink is relative to directory of argument. Replace the
+          // symlink with the resolved name in the same directory.
           const size_t p_len = strlen(p);
           const size_t buf_len = strlen(buf);
-          cpy = xmalloc(p_len + buf_len + 1);
-          memcpy(cpy, p, p_len);
-          memcpy(path_tail((char_u *)cpy), buf, buf_len + 1);
-          xfree(p);
-          p = cpy;
+          p = xrealloc(p, p_len + buf_len + 1);
+          memcpy(path_tail((char_u *)p), buf, buf_len + 1);
         } else {
           xfree(p);
           p = xstrdup(buf);

